### ISSUE TITLE
Allow Graphing SQLAlchemy relationships

### ISSUE THREAD

**Issue by kassett** [sqlalchemy/sqlalchemy#12201]
### Describe the use case

Support building a directional graph capable of iterating over relationships and describing the path between relationships. I have actually already written and am using code similar to this. One downside is that it obfuscates how potentially slow iterating over many relationships is -- however, there are potentially cases where iterating over many relationships could be faster than doing complicated joins. This might be true, for instances, when using the MyRocks plugin for MySQL. 
It also might be nice to be able to visualize the relationships between objects in your codebase.
I have a PR ready to be opened if this is a feature that you're interested in.

### Databases / Backends / Drivers targeted

N/A / All

### Example Use

```
class Parent(Base):
    __tablename__ = "parents"
    id = Column(Integer)
    children = relationship("Child")

class Child(Base):
    __tablename__ = "children"
    id = Column(Integer)
    parent_id = Column(Integer, ForeignKey("parents.id")
    parent = relationship("Parent")

graph = RelationshipGraph(Base)
path = graph.get_path(Child, Parent)
assert path == ["parent"]

child = session.get(Child, 55)
parent = graph.traverse(child, Parent)
```

**Comment by CaselIT** [comment]
Hi,

If I'm not mistaken, some of this capability is already present since it's used internally by the orm.

That said for such features usually the preferred way would be to release it as an external library.

Does it use any private api that may make sense consider to make public?

**Comment by kassett** [comment]
Thanks for the quick response @CaselIT . Happy holidays BTW. If I'm understanding what you're saying correctly, you're asking if there are any private functions right now that could expose the same functionality simply by making them public? I don't think there are, but such code is only about 200 lines, so making an entirely new library seems like kind of a waste. Would a code snippet help, so you could see exactly what features I'm using?

**Comment by CaselIT** [comment]
>you're asking if there are any private functions right now that could expose the same functionality simply by making them public?

No, I'm asking if your code uses some private functionality that you complicate RelationshipGraph from being an external library.

> such code is only about 200 lines, so making an entirely new library seems like kind of a waste. 

I understand that it may not seem much, but there are many features that could be added that are small by themself, but amount to a large maintenance burden for the sqlalchemy team. That's why we prefer the external library route for what can work either way.

@zzzeek do you have some input here?

**Comment by kassett** [comment]
So yes, this code would add networkx as an additional dependency -- albeit presumably an optional one. 

I understand that small amounts code can be difficult to maintain, but as long as this code relies on the internals of the inspect module, I don't think you need to worry too much about maintainability. 

Additionally, I proposed this change as a nice-to-have, but I would like to stress that in its current state, maintaining the declaration of models and their relationships to one another can be exceedingly complicated. This code could make it useful to describe your tables and visualize their relationships to one another, as a form of sanity checking.

**Comment by CaselIT** [comment]
That information is given by the uselist flag on the many side. I think the reason it's missing is because it's not necessarily a given that only cases where unique fk are present should be considered as one to one

**Comment by CaselIT** [comment]
Also could you detail what would this feature do, and what do you find excitingly complicated woth the current api?

**Comment by kassett** [comment]
I envision a class with four key methods: `traverse`, `has_path`, `get_path`, and `visualize`:

1. **`has_path`**: Returns a boolean indicating whether a specific sequence of relationships exists to allow traversal from one table to another.
2. **`get_path`**: Provides a list of attributes representing the traversal path between two tables (e.g., `parent -> grandparent`).
3. **`traverse`**: Takes an object instance and performs the traversal to another object. For example, `graph.traverse_to(child, Grandparent)` would return an instance of `Grandparent`.
4. **`visualize`**: Generates a graph of ORM-defined relationships, where edges are labeled with the attribute enabling traversal and the type of relationship. A textual representation could also be used.

For example:

![example-graph](https://github.com/user-attachments/assets/b2f4e2b5-11d4-4691-a227-86c33d577ccc)

### Why this is beneficial:
Currently, it is challenging to write tests that enforce specific ORM structures, especially with more than five relationships per object. This can lead to confusion and difficulty maintaining consistency. A utility like this would enable tests like:

```python
assert RelationshipGraph(Base).has_path(Child, Grandparent)
```

**Comment by zzzeek** [comment]
SQLAlchemy's mapping relationships are fully [introspectable](https://docs.sqlalchemy.org/en/20/orm/mapping_styles.html#orm-mapper-inspection-mapper) and there are already many tools which do this:   https://pypi.org/project/sqlalchemy-schemadisplay/ https://pypi.org/project/SQLAlchemyViz/   take a look at those
